docs/understanding-airbyte/airbyte-protocol.md

```diff
@@ -0,0 +1,735 @@
+
+The key primitives that the Protocol uses to describe data are Catalog, Configured Catalog, Stream, Configured Stream, and Field:
+* Stream - A Stream describes the schema of a resource and various metadata about how a user can interact with that resource. A resource in this context might refer to a database table, a resource in a REST API, or a data stream.
+* Field - A Field refers to a "column" in a Stream. In a database this would be a column; in a JSON object it is a field.
+* Catalog - A Catalog is a list of Streams that describes the data in the data store that a Source represents.
+
+An Actor can advertise information about itself with an [Actor Specification](#actor-specification). One of the main pieces of information the specification shares is what information is needed to configure an Actor.
+
+Each of these concepts is described in greater depth in their respective section.
+
+## Actor Interface
+This section describes important details about the interface over actors. It reviews parts of the interface that are the same across all actors. It also describes some invariants for all methods in actor interfaces.
+
+### Common Interface Methods
+The following part of the interface is identical across all actors:
+```
+spec() -> ConnectorSpecification
+check(Config) -> AirbyteConnectionStatus
+```
+
+These methods are described in their respective sections ([spec](#spec), [check](#check)).
+
+### Interface Invariants
+
+The output of each method in actor interface is wrapped in an `AirbyteMessage`. This struct is an envelope for the return value of any message in the described interface. See the section the [AirbyteMessage](#airbytemessage) section below for more details. For the sake of brevity, interface diagrams will elide these `AirbyteMessage`s.
+
+Additionally, all methods described in the protocol can emit `AirbyteLogMessage`s and `AirbyteTraceMessage`s (for more details see [Logging](#airbytelogmessage)). These messages allow an actor to emit logs and other informational metadata. All subsequent method signatures will assume that any number of messages of these types (wrapped in the `AirbyteMessage`) may be emitted.
+
+Each method in the protocol has 3 parts:
+1. **Input**: these are the arguments passed to the method.
+2. **Data Channel Egress (Output)**: all outputs from a method are via STDOUT. While some method signatures declare a single return value, in practice, any number of `AirbyteLogMessage`s and `AirbyteTraceMessage`s may be emitted. An actor is responsible for closing STDOUT to declare that it is done.
```

Comment:
... No connector should emit any data on STDERR.

Reply:
i don't follow this comment. is the clarification that STDERR is only ever used for error logs? everything else happens on STDOUT?

Reply:
added a line about this in the `Additional Invariants` section